File: app/domain/status_machine.py
"""Motor de estados configurable por tenant.

Cada restaurante define sus propios estados y transiciones validas en
las tablas order_statuses / order_status_transitions. Este modulo solo
valida contra esa configuracion: no conoce ningun estado por nombre.
"""

from dataclasses import dataclass

# Categorias normalizadas. El nombre del estado varia por restaurante,
# la categoria no: el KDS y los reportes se apoyan en ella.
CATEGORIES = ("new", "kitchen", "ready", "in_transit", "completed", "cancelled")


@dataclass(frozen=True)
class Status:
    id: str
    code: str
    category: str
    is_initial: bool
    is_final: bool


@dataclass(frozen=True)
class Transition:
    from_status_id: str
    to_status_id: str
    required_permission: str | None


class TransitionError(Exception):
    pass


class StatusMachine:
    def __init__(self, statuses: list[Status], transitions: list[Transition]):
        self._statuses = {s.id: s for s in statuses}
        self._transitions = {
            (t.from_status_id, t.to_status_id): t for t in transitions
        }

    def initial(self) -> Status:
        for s in self._statuses.values():
            if s.is_initial:
                return s
        raise TransitionError("El tenant no tiene estado inicial configurado")

    def allowed_from(self, status_id: str) -> list[Status]:
        return [
            self._statuses[to_id]
            for (frm, to_id) in self._transitions
            if frm == status_id
        ]

    def validate(self, from_id: str, to_id: str, permissions: list[str]) -> None:
        if self._statuses[from_id].is_final:
            raise TransitionError("El pedido ya está en un estado final")

        transition = self._transitions.get((from_id, to_id))
        if transition is None:
            raise TransitionError("Transición no permitida para este restaurante")

        needed = transition.required_permission
        if needed and needed not in permissions:
            raise TransitionError(f"Falta el permiso: {needed}")

File: php/src/Domain/StatusMachine.php
<?php

declare(strict_types=1);

namespace App\Domain;

/**
 * Motor de estados configurable por tenant.
 *
 * Cada restaurante define sus propios estados y transiciones validas en las
 * tablas order_statuses / order_status_transitions. Esta clase solo valida
 * contra esa configuracion: no conoce ningun estado por nombre.
 */
final class StatusMachine
{
    /** @var array<string, OrderStatus> */
    private array $statuses = [];

    /** @var array<string, StatusTransition> */
    private array $transitions = [];

    /**
     * @param OrderStatus[] $statuses
     * @param StatusTransition[] $transitions
     */
    public function __construct(array $statuses, array $transitions)
    {
        foreach ($statuses as $s) {
            $this->statuses[$s->id] = $s;
        }
        foreach ($transitions as $t) {
            $this->transitions[$this->key($t->fromStatusId, $t->toStatusId)] = $t;
        }
    }

    private function key(string $fromId, string $toId): string
    {
        return "{$fromId}>{$toId}";
    }

    public function initial(): OrderStatus
    {
        foreach ($this->statuses as $s) {
            if ($s->isInitial) {
                return $s;
            }
        }
        throw new TransitionError('El tenant no tiene estado inicial configurado');
    }

    /** @return OrderStatus[] */
    public function allowedFrom(string $statusId): array
    {
        $result = [];
        foreach ($this->transitions as $t) {
            if ($t->fromStatusId === $statusId) {
                $result[] = $this->statuses[$t->toStatusId];
            }
        }
        return $result;
    }

    /** @param string[] $permissions */
    public function validate(string $fromId, string $toId, array $permissions): void
    {
        if ($this->statuses[$fromId]->isFinal) {
            throw new TransitionError('El pedido ya esta en un estado final');
        }

        $transition = $this->transitions[$this->key($fromId, $toId)] ?? null;
        if ($transition === null) {
            throw new TransitionError('Transicion no permitida para este restaurante');
        }

        $needed = $transition->requiredPermission;
        if ($needed !== null && !in_array($needed, $permissions, true)) {
            throw new TransitionError("Falta el permiso: {$needed}");
        }
    }
}
